### ISSUE TITLE
Ability to stop on unhandled error

### ISSUE THREAD

**Issue by fatcerberus** [svaarala/duktape#230]
The Duktape debugger should automatically stop at the point of throw for unhandled exceptions, similar to how native debuggers stop when an app would otherwise segfault.

**Comment by svaarala** [comment]
Agree, that's part of identified future work: https://github.com/svaarala/duktape/blob/master/doc/debugger.rst#more-flexible-pausing

**Comment by fatcerberus** [comment]
As a sort of workaround, if I were to hypothetically put a `debugger;` statement in `Duktape.errThrow()`, would that get me what I want?  Or has the stack already been unwound by that point?

**Comment by svaarala** [comment]
It wouldn't be unwound, but it would break even when the error was not an uncaught one.

**Comment by fatcerberus** [comment]
Is there a duktape API equivalent to the `debugger` statement?  Now that I think of it my errCreate handler is implemented in C so that's where I would have to trigger the breakpoint.

**Comment by svaarala** [comment]
@fatcerberus Not right now, that's probably something to add to the API. It should be very easy to implement.

Another API call I'd like to have is the ability to send application specific notifications (statistics etc) which can be shown generically in the debug client (or handled specially if the debug client understands them).

**Comment by fatcerberus** [comment]
@svaarala Hm, so I was going to try my hand at implementing this feature myself.  From what I could tell, changes would need to be made to 2 files: `duk_error_longjmp,c` and `duk_debugger.c`.

My idea was to change it so that if there is no `jmpbuf_ptr` when an error is thrown and the debugger is attached, to trigger a debug pause instead of panicking.  But then I realized: This won't work in the case of a top-level protected call, since as far as Duktape is concerned at that point, the error is actually being handled.  Since I use `duk_pcall()` to call the initial `game()` function, such a change wouldn't actually get me what I want.

**Comment by svaarala** [comment]
Perhaps the condition could be that no Ecmascript code is catching the error - but then again, the top level catcher could also be:

    duk_eval_string_noresult(ctx, "try { myMainLoop(); } catch (e) { print(e.stack || e); }");

**Comment by fatcerberus** [comment]
That's true, if a strange way to implement a bootstrap.  I think I'd be fine with a break-if-no-JS-catch rule, the mentioned caveat would be a fair tradeoff to be able to use the debugger.

Do catchstack entries differentiate between JS `catch` and protected C calls currently, or would that have to be implemented as well?

**Comment by svaarala** [comment]
Assuming that bootstrap was only called once, it'd be a very natural bootstrap for me at least :-)

It's short and handles the error printing better than most C call sites would, and since it's almost never called, its performance doesn't matter.

**Comment by svaarala** [comment]
Catchstack entries are only used for Ecmascript try-catch etc. There's essentially no record for e.g. duk_safe_call().

Also, for C code it might not be all that rare for a catch site to do cleanup and then rethrow, essentially doing what "finally" does.  Duktape has no knowledge what's going to happen after a catch, though.

Pausing on "uncaught error" is pretty clearly useful when it is defined in terms of "user code", say a web page, so that it discounts whatever "infrastructure code" is present. However, the difficulty is that not all infrastructure code is Duktape code. For example, the user may have a Promise framework which does catch/rethrow for errors as a normal part of its operation. Those catch sites would need to be ignored in some way when deciding on whether or not to pause.

It'd be much easier to pause on any throw :-)

**Comment by svaarala** [comment]
One approach to this would be to rely on the debug client: pause on every error throw, and let the debug client inspect the call stack and other state to decide whether to autoresume without user interaction or not.

As things are right now, the debug client probably still wouldn't have enough information to make that call correctly. It could discount any call sites in specific files ("infrastructure code") but it couldn't decide what to make of a native call site, even if it knew there was one: native catch sites are not registered the same way and the client would need to know e.g. the filename where the catch site occurs to identify it as an "infrastructure" site.

Perhaps "pause on error not caught by current function" or "pause on error not caught in current file" would work reasonably? (It's not the same for many use cases, of course.)

**Comment by fatcerberus** [comment]
Pause on all might be preferable in fact.  It would allow the client to pause on first-chance exceptions as well if the user so desired, a use case which is sometimes helpful.

There is one potential issue here: Pausing on throw breaks the contract that throwing an error doesn't return if the user resumes afterward.  Perhaps it might be better to pause during error augmentation instead, but then it would be impossible to differentiate between first and second-chance exceptions...

This is a deceptively thorny issue! :)

**Comment by svaarala** [comment]
It's a thorny issue indeed, which is why it was not included in the first debugger implementation phase.

Btw, note that you can e.g.:

    throw 1;

in which case no error is constructed.

**Comment by fatcerberus** [comment]
Huh.  That's quite distinct from, e.g. C# which only allows you to throw Exception instances or derivations thereof.

**Comment by fatcerberus** [comment]
One way I think this could be done would be to implement it as part of processing for the THROW opcode.  This wouldn't catch errors thrown by native code (`duk_require_xx()`, `duk_error()`), but it would at least allow consistency to be maintained: When a throw is encountered without an active catch, the bytecode executor could just pretend it just hit a breakpoint.

This wouldn't break the no-return contract since resuming would simply execute the THROW again.  Of course this is all assuming that THROW has no side effects other than throwing an error...

**Comment by fatcerberus** [comment]
So I added this line to the top of `case DUK_EXTRAOP_THROW` (throw is an "extra" opcode? Seems like that would have made the shortlist...):

```c
#if defined(DUK_USE_DEBUGGER_SUPPORT)
				DUK_D(DUK_DPRINT("THROW with attached debugger, halt execution"));
				if (DUK_HEAP_IS_DEBUGGER_ATTACHED(thr->heap)) {
					DUK_HEAP_SET_PAUSED(thr->heap);
					--act->pc;  /* back up so we pause just before throwing */
					goto restart_execution;
				}
#endif
```

It works, but pauses on any thrown error, as I couldn't find a way to limit it to unhandled ones (see next paragraph).  Except resuming is now impossible; due to the PC decrement, it just keeps hitting the THROW and pausing again.  So I'm thinking the error object needs to be augmented in this case, so that when the THROW is encountered a second time (on resume), it actually throws instead of pausing.

Regarding the issue of unhandled errors: I discovered this assert in THROW opcode handling:
```c
DUK_ASSERT(thr->heap->lj.jmpbuf_ptr != NULL);  /* always in executor */
```

How can I tell if an error is unhandled, if there's always a valid catchpoint in this context?

**Comment by fatcerberus** [comment]
Okay, so I solved the issue of unresumability by adding a new opcode, `DUK_EXTRAOP_THROWU` (throw unconditionally).  A throw statement emits both a THROW and a THROWU (the second is only emitted if debugging is enabled); the first triggers the debugger pause, the second will throw the error anyway.  In this way you can still continue and unwind the stack if needed.

**Comment by fatcerberus** [comment]
@svaarala Do you mind if I open a pull request for this once I get everything working? :)

**Comment by svaarala** [comment]
This has been fixed by #286.
